### DISCUSSION TITLE
io.netty.handler.codec.DefaultHeaders.remove(DefaultHeaders.java 1074)   java.lang.NullPointerException : Cannot read field ""hash"" because ""e"" is null

### DISCUSSION THREAD

**Discussion by jeffcf** [netty/netty#14159] [Q&A]
I have a question and I want to ask an expert.
When I use SpringBoot webflux, I sometimes get the following error:
--------------------------------------------------------------
Exception  :
io.netty.handler.codec.DefaultHeaders.remove0(DefaultHeaders.java 1074)   java.lang.NullPointerException : Cannot read field ""hash"" because ""e"" is null 

![image](https://github.com/netty/netty/assets/17584661/4ac35cc2-a86c-46b0-aba8-1a402390a8dc)

-------------------------------------------------------------------------------------------------------
Env:
SpringBoot 3.2 with WebFlux  
netty-codec-4.1.109Final.jar
-------------------------------------------------------------------------------------------------------
DefaultHeaders.java SourceCode

    /**
     * @return the first value inserted whose hash code equals {@code h} and whose name is equal to {@code name}.
     */
    private V remove0(int h, int i, K name) {
        HeaderEntry<K, V> e = entries[i];
        if (e == null) {
            return null;
        }

        V value = null;
        HeaderEntry<K, V> next = e.next;
        while (next != null) {
            if (next.hash == h && hashingStrategy.equals(name, next.key)) {
                value = next.value;
                e.next = next.next;
                next.remove();
                --size;
            } else {
                e = next;
            }

            next = e.next;
        }

        e = entries[i];
        if (e.hash == h && hashingStrategy.equals(name, e.key)) {      // <-------------------1074 Exception Line
            if (value == null) {
                value = e.value;
            }
            entries[i] = e.next;
            e.remove();
            --size;
        }

        return value;
    }
-------------------------------------------------------------------------------------------------------

It's hard for me to reproduce this error, it happens occasionally

Has anyone encountered this error? 
thank you all.

**Comment by jeffcf** [comment]
Now there is a new discovery that this bug can be reproduced when the http header has Except:100-continue.

Use .net http client or jmeter to perform a stress test. Netty NPE errors will appear in about 2-5 minutes.

By default, the ".net" http client will contain Except: in the http header Except:100-continue.

( PS: .net core does not have this http header by default)

**Comment by normanmaurer** [comment]
I wonder if you might send the same `HttpObject` down the pipeline without duplicating it ?

**Comment by jeffcf** [comment]
Hi normanmaurer,

The two methods testing are as follows

1. JMeter --> .net http api --> netty server
2. JMeter --> netty server

Http Object should be unique.

The same test method,when I remove Except:100-Contine, this error will not occur.

Thanks

**Comment by normanmaurer** [comment]
@violetagg can you check... for me it looks like the same object might be shared across threads.

**Comment by violetagg** [comment]
@jeffcf @normanmaurer Without some example code it is hard to say. We have fixes related to `Except:100-Continue` handling, but here it is not obvious the version of Reactor Netty.

**Comment by jeffcf** [comment]
@violetagg 

SpringBoot 3.2 with WebFlux
netty-codec-4.1.109Final.jar


io.netty.handler.codec.DefaultHeaders.remove0(DefaultHeaders.java 1074) java.lang.NullPointerException : Cannot read field ""hash"" because ""e"" is null

pom.xml
![image](https://github.com/user-attachments/assets/892b397e-6174-4230-bf22-dbedd45c7331)




maven depency 
![image](https://github.com/user-attachments/assets/794aaeaa-a8a8-4995-8bac-a0e1df356100)

![image](https://github.com/user-attachments/assets/db8610a2-63aa-47a2-82da-bb4e7aeecae1)

![image](https://github.com/user-attachments/assets/6664f8ce-3bbd-45b7-aa84-74b5d7d1b2ac)

Thanks

**Reply by violetagg** [reply]
Spring Boot 3.2.5 comes with Reactor Netty 1.1.18 (io.projectreactor.netty:reactor-netty-http) which mean you have all fixes related to `Expect`. Can you provide some example code?

**Reply by jeffcf** [reply]
@violetagg

If I don't need the Except 100-continue feature, is there a way to turn it off directly and avoid this error?

I retested and made my example code  simpler.

sample code as attachments (Java Maven Project)

[new-netty-gw.zip](https://github.com/user-attachments/files/16265905/new-netty-gw.zip)


-----------------------------------------
startup springboot netty-gw instances
1. 8080 port for test endpoint http://127.0.0.1:8080/rest  (update application.yml   server:port:8080)

![image](https://github.com/user-attachments/assets/cf8355db-115d-4491-bab4-375a863c9363)

-----------------------------------------
use jmeter  call http://127.0.0.1:8080/rest
-----------------------------------------
http header  
--> Expect: 100-continue

![image](https://github.com/user-attachments/assets/bed5eb97-c7cc-4d64-936a-f1723ee1da0b)

-----------------------------------------
http body
{
 "HEADER"	 : {
 	"APID": "XX",
 	"APID": "XX"
 },
 "BODY" : {
 	"BOOK_NAME" : "test"
 }
}
![image](https://github.com/user-attachments/assets/896328ec-8c60-4c1f-b86e-1c836af7d126)

-----------------------------------------
error log

![image](https://github.com/user-attachments/assets/969c74a3-3476-4a67-9a50-abf2806395ed)


Thanks

**Comment by jeffcf** [comment]
@violetagg 
I'm sorry to bother you. I have attached the test code to the discussion stream. I would like your help to see if I can find the problem.

thank you

**Reply by violetagg** [reply]
@jeffcf What are your exact configurations for the `Thread Group`?

**Reply by violetagg** [reply]
Actually I might have an idea ... Can you open an issue here https://github.com/reactor/reactor-netty/issues

**Reply by jeffcf** [reply]
@violetagg 
What are your exact configurations for the Thread Group?
----> Default value, no special setting Thread Group

Actually I might have an idea ... Can you open an issue here https://github.com/reactor/reactor-netty/issues
---> OK
---> https://github.com/reactor/reactor-netty/issues/3363

**Comment by violetagg** [comment]
@normanmaurer We are sure that the object is new (no reuse), in the picture below (provided by @jeffcf),

![351252767-29e0564a-1da0-4eed-b5c3-d74d209d6263](https://github.com/user-attachments/assets/aa3a5b69-c6e1-438c-9123-8395e9d7f059)

**Reply by violetagg** [reply]
@normanmaurer Please ignore the comment above. It seems that the fix in Reactor Netty is working as expected.

**Comment by normanmaurer** [comment]
Have an link ?Am 23.07.2024 um 08:29 schrieb Violeta Georgieva ***@***.***>:﻿
@normanmaurer Please ignore the comment above. It seems that the fix in Reactor Netty is working as expected.

**Reply by violetagg** [reply]
https://github.com/reactor/reactor-netty/issues/3363
